### ISSUE TITLE
Weird HTML formatting

### ISSUE THREAD

**Issue by lbogdan** [prettier/prettier#7153]
**Prettier 1.19.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeAJgSwG4D4A6UABEagIQC0FRA7nJgE7pFUHGkMQCu8DFANpigBrIkhgQAvPhDAiUAIYBbOEiIByLgGc4DdQBoiABwUNlWtXMzo11ogGoiARiIBfNzNwA5OAA8YRACqOgyoAPScPLoCQsJsJORURBDChgBGPLSYMAAWRH4wZrQ52XBaJmBwLBTxHNy8MSJiEtKy8sqqGtq6Bsam5pZE1rbMji7urp6EJCQ+-kEh06QR9dGCImzhWHiEIPogEEYwmNBayKCmnDQACqYIZygK2BDWeyBpZmDCcDAAyhVCADmyEKXDg+xyMCU-AA6iV4OUFJVfvdsjhsgBPZDgLRnfZCEIwa5mQFKBTIABmCn4On2ACstL4AEKfb5-DoAGSEcEp1NpIAZvl+QP4cAAilwIPBeTTwSATAwQtjIdC3kYGEIYDDrLlkAA2AAM+3VEB0MLMRmx6rKumwPP2AEdJfBiYcHiAFFoKFA4HB0H63gw4E7GHBiQpSeSkFTZfsdEpMCCGGC4yLxc6edG+XKYAo0tr0LqkAAmfaFBSYdaAgDCECUZOxUGg9pA3QAKnmHjH+dgwQBJKD+2C-MAao4AQUHvxgGNFMp0rlcQA)
```sh
# Options (if any):
--parser html
--print-width 60
```

**Input:**
```jsx
<div>
  <router-link :to="{ name: 'user', params: { id: id + 1 } }">Next User</router-link>
</div>
```

**Output:**
```jsx
<div>
  <router-link
    :to="{ name: 'user', params: { id: id + 1 } }"
    >Next User</router-link
  >
</div>
```

**Expected behavior:**

Either

```jsx
<div>
  <router-link
    :to="{ name: 'user', params: { id: id + 1 } }"
  >
    Next User
  </router-link>
</div>
```

or (even **better**, IMO, as it keeps the whitespace consistent)

```jsx
<div>
  <router-link
    :to="{ name: 'user', params: { id: id + 1 } }"
  >Next User</router-link>
</div>
```

**Comment by tdeea** [comment]
Having the same issue and it`s very annoying. Congrats for the initiative.
I would personally like and output like this:
```
<div>
  <router-link
      :to="{ name: 'user', params: { id: id + 1 } }">
       Next User
  </router-link>
</div>
```

**Comment by thorn0** [comment]
@lbogdan Please make the title of this issue more descriptive, so that it was easier to find it when somebody decides to look into improving this.

**Comment by evilebottnawi** [comment]
```
<div>
  <router-link
    :to="{ name: 'user', params: { id: id + 1 } }"
  >
    Next User
  </router-link>
</div>
```

inserting newlines and spaces can potential break html (html like), i agree what in your case it is better, but it is not safe in other cases

**Comment by lbogdan** [comment]
@tdeea While formatting is usually subjective 🙂 , I think there's two issues with yours - it still looks weird to me when:
- keeping whitespace
```html
<div>
  <router-link
      :to="{ name: 'user', params: { id: id + 1 } }">Next User</router-link>
</div>
```
- you have long lines and only focus on the beginning of the line:
```html
<div>
  <router-link
      :to="{ name: 'user', params: { id: id + 1 } }" ................................................................................................................................................>
       Next User
  </router-link>
</div>
```

@thorn0 I though of this, but couldn't come up with a short, descriptive title for this. That is to say I'm open to suggestions!

@evilebottnawi Totally agree, that's why I said my second version was better, because it preserves whitespaces (or better said, the absence of it).

**Comment by thorn0** [comment]
@lbogdan The problem with the second version is the content that can come immediately after the closing tag. What if it is another inline element with a lot of attributes? How would that look?

**Comment by lbogdan** [comment]
@thorn0 Good point, that indeed looks better in your case, hopefully you'll agree with me that without a following element it looks weird. Maybe we can differentiate between the two?
Also, shouldn't the `>` of the opening tag be aligned with the `<`? Is that a bug?
```diff
 <div>
   <router-link
     :to="{ name: 'user', params: { id: id + 1 } }"
-    >Next User</router-link
+  >Next User</route-link
   ><router-link
     :to="{ name: 'user', params: { id: id + 1 } }"
-    >Next User</router-link
+  >Next User</router-link
   >
 </div>
```

**Comment by thorn0** [comment]
> Maybe we can differentiate between the two?

We can. There are other known special cases that can get special handling too (e.g. it's safe to add whitespace to an inline tag if it's the only child of a block tag). My personal concern here is that the result might turn out too inconsistent and even more unexpected. E.g. you add a tag and its preceding tag gets suddenly reformatted.

> Also, shouldn't the > of the opening tag be aligned with the <? Is that a bug?

No. Indenting the content of the tag is in intentional.

**Comment by lbogdan** [comment]
> No. Indenting the content of the tag is intentional.

@thorn0 Hm, but isn't that inconsistent with the whitespace version?

```html
<div>
  <router-link
    :to="{ name: 'user', params: { id: id + 1 } }"
  >
    Next User
  </router-link>
</div>
```

**Comment by thorn0** [comment]
@lbogdan To me, it's not. For both versions, the content is indented.

**Comment by lbogdan** [comment]
Going back to your previous example, why couldn't it look like this?

```html
<div>
  <router-link
    :to="{ name: 'user', params: { id: id + 1 } }"
  >Next User</router-link>
  <router-link
    :to="{ name: 'user', params: { id: id + 1 } }"
  >Next User</router-link>
</div>
```

**Comment by thorn0** [comment]
You added whitespace between the tags.

**Comment by lbogdan** [comment]
Fair enough. Still, assuming the whitespace is already there, is still gets reformatted to your version, which still feels sub-par.

**Comment by thorn0** [comment]
The indentation of the content is needed to show the hierarchy. Content isn't always this simple, it might contain nested things.

**Comment by thorn0** [comment]
I didn't say there is no room for improvement.

**Comment by lbogdan** [comment]
While I understand that's a very complex subject, it still feels to me that you're optimizing / thinking about more complex use-cases, leaving these simpler ones (which I'd say happen more often, but that's just a guess) looking weird enough to be quite annoying, and maybe for some even making them quit using prettier to format HTML.

That being said, let me know if / how I can help in improving this!

**Comment by thorn0** [comment]
It's not much about complex vs simple use-cases as it's about general case vs special cases. The problem first of all should've been solved for the general case, which was done: HTML formatting was [implemented a year ago](https://prettier.io/blog/2018/11/07/1.15.0.html#htmlvueangular) by @ikatyang 🙏. Since then, this implementation hasn't changed much and there has been very few complaints that it broke someone's code, which is the main priority for Prettier. However, this is a general-case solution and it's kind of still in its early stage.
